File: example/typescript/class/factory-constructor.ts
class Logger {
  static readonly _cache: Map<string, Logger> = new Map();

  static getInstance(name: string) {
    if (this._cache.has(name)) {
      return this._cache.get(name)
    } else {
      const logger = new Logger(name)
      this._cache.set(name, logger)
      return logger
    }
  }

  constructor(public name: string){}
}

const logger1 = Logger.getInstance('smsLogger')
const logger2 = Logger.getInstance('smsLogger')
console.log(logger1 == logger2);  // true

const logger3 = Logger.getInstance('mobiLogger')
console.log(logger1 == logger3);  // false

File: example/dart/class/factory-constructor.dart
class Logger {
  final String name;
  static final Map<String, Logger> _cache = <String, Logger>{};

  factory Logger(String name) {
    if (_cache.containsKey(name)) {
      return _cache[name];
    } else {
      final logger = new Logger._internal(name);
      _cache[name] = logger;
      return logger;
    }
  }

  Logger._internal(this.name);
}

main(List<String> args) {
  var logger1 = new Logger('smsLogger');
  var logger2 = new Logger('smsLogger');
  print(logger1 == logger2);  // true

  var logger3 = new Logger('mobiLogger');
  print(logger1 == logger3);  // false
}